core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -77,7 +77,7 @@ public void accept(String file) {
   private final String commitUUID = UUID.randomUUID().toString();
   private final AtomicInteger attempt = new AtomicInteger(0);
   private final List<String> manifestLists = Lists.newArrayList();
-  private Long snapshotId = null;
+  private volatile Long snapshotId = null;
```

Comment:
The volatile keyword is needed as described [here](https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html). Credits to @raptond for the link.